packages/next/src/server/web/spec-extension/unstable-cache.ts

```diff
@@ -239,33 +239,72 @@ export function unstable_cache<T extends Callback>(
-                        `revalidating cache with key: ${invocationKey}`,
-                        err
-                      )
-                    )
+                if (workStore.isRevalidate) {
+                  // When the page is revalidating and the cache entry is stale,
+                  // we need to wait for fresh data (blocking revalidate)
+                  let pendingRevalidate =
+                    workStore.pendingRevalidates[invocationKey]
+
+                  if (!pendingRevalidate) {
+                    // Create the revalidation promise
+                    pendingRevalidate = workUnitAsyncStorage
+                      .run(innerCacheStore, cb, ...args)
+                      .then(async (result) => {
+                        await cacheNewResult(
+                          result,
+                          incrementalCache,
+                          cacheKey,
+                          tags,
+                          options.revalidate,
+                          fetchIdx,
+                          fetchUrl
+                        )
+                        return result
+                      })
+
+                    // Attach empty catch to prevent unhandled rejection warnings
+                    // This mirrors handling in patch-fetch
+                    pendingRevalidate.catch(() => {})
+
+                    workStore.pendingRevalidates[invocationKey] =
+                      pendingRevalidate
+                  }
+
+                  // Wait for and return the revalidated result
+                  return pendingRevalidate
+                } else {
+                  // In App Router we return the stale result and revalidate in the background
+
+                  // Check if there's already a pending revalidation to avoid duplicate work
+                  if (!workStore.pendingRevalidates[invocationKey]) {
+                    // We run the cache function asynchronously and save the result when it completes
+                    workStore.pendingRevalidates[invocationKey] =
+                      workUnitAsyncStorage
+                        .run(innerCacheStore, cb, ...args)
+                        .then((result) => {
+                          return cacheNewResult(
+                            result,
+                            incrementalCache,
+                            cacheKey,
+                            tags,
+                            options.revalidate,
+                            fetchIdx,
+                            fetchUrl
+                          )
+                        })
+                        // @TODO This error handling seems wrong. We swallow the error?
+                        .catch((err) =>
+                          console.error(
+                            `revalidating cache with key: ${invocationKey}`,
+                            err
+                          )
+                        )
+                  }
+                }
```

Comment:
Do we need to branch the whole `pendingRevalidates` logic instead of an `if (workStore.isRevalidate) return workStore.pendingRevalidate[invocationKey])`?